Experimental: JS compression via Closure Compiler

rr-it opened the pull request:
### Description:

Experimental: This introduces JS compression of `piwik.js` via `Closure Compiler`.

See https://github.com/matomo-org/matomo/issues/20964

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

rr-it committed:
Experimental: add Closure Compiler to build-tracker-js with `SIMPLE_OPTIMIZATIONS`

rr-it committed:
Fix: Closure Compiler warning `JSC_USELESS_CODE` in `for`-loops

```
var i = 0;
for (i; i < object.length; i++) {}
```

becomes

```
var i;
for (i = 0; i < object.length; i++) {}
```

Alternative (unreadable):

```
var i = 0;
for (; i < object.length; i++) {}
```

rr-it committed:
Fix: Closure Compiler warning `JSC_MISPLACED_ANNOTATION`

rr-it committed:
Experimental: use Closure Compiler with `ADVANCED_OPTIMIZATIONS`

rr-it committed:
Fix: Closure Compiler warning `JSC_TYPE_PARSE_ERROR` - Bad type annotation in JSDoc

rr-it committed:
Fix: Closure Compiler error `JSC_UNDEFINED_VARIABLE` - `Matomo_Overlay_Client` and `piwik_track` undeclared

rr-it committed:
Experimental: suppress JSDoc type checks in `piwik.js` for Closure Compiler

This disables type warnings: `JSC_UNRECOGNIZED_TYPE_ERROR`, `JSC_TYPE_MISMATCH`, `JSC_WRONG_ARGUMENT_COUNT` and more.

rr-it committed:
Experimental: add `externs` base definition for `piwik.js` used by Closure Compiler

rr-it committed:
Experimental: add deprecated functions to `externs` definition for `piwik.js`

rr-it committed:
Experimental: add global methods used by Matomo to `externs` definition for `piwik.js`

rr-it committed:
Experimental: add public Matomo API to `externs` definition for `piwik.js`

rr-it committed:
Experimental: add public Tracker API methods to `externs` definition for `piwik.js`

+ Copy all functions that begin with `this.nameOfFunction` from `function Tracker()` in `piwik.js`. Exclude functions inside `/*<DEBUG>*/ … /*</DEBUG>*/`-markers.
+ Rename inside `piwik.externs.js` via find `^this\.` and replace by `Tracker.`
+ Empty method bodies inside `piwik.externs.js` via find `\{(\n .+|\n)+\n\};` and replace by `{};`

rr-it committed:
Experimental: add public Tracker API methods to `externs` definition for `piwik.js` - fix `Tracker.optUserOut`

rr-it committed:
JSDoc: declare constructors

rr-it committed:
JSDoc: getAsyncTracker might return no Tracker at all

rr-it committed:
JSDoc: mark `piwik_log` and `piwik_track` as deprecated

rr-it committed:
Experimental: enclose fixed property names of data-object in quotes to exclude them from Closure Compiler munging

rr-it committed:
Build tracker js

rr-it committed:
Experimental: add `DEBUG` variable for Closure Compiler

bx80 commented:
Thanks for this analysis and contribution @rr-it, it's much appreciated :+1: We'll review this PR and schedule some testing.

A ~20% decrease in tracking code size sounds great, even if only a few kilobyte, it can soon add up to a useful traffic reduction when serving many millions of requests per month.

rr-it committed:
Experimental: enclose still missing fixed property names of data-object in quotes to exclude them from Closure Compiler munging

rr-it committed:
Experimental: prepend all private property names with an underscore

Some common JS names defined in the default Closure Compiler `externs` are not munged e.g.:
location, name, target, type, href, enabled, timeout, interval

When renamed by prepending with an underscore they get correctly munged.

rr-it committed:
Experimental: rename functions colliding with common names

Some common JS names defined in the default Closure Compiler `externs` are not munged:
find, push, trim

When renamed the functions and calls get correctly munged.

rr-it committed:
Experimental: add missing global methods used by Matomo to `externs` definition for `piwik.js`

rr-it committed:
Build tracker js
